docs/design/mesos-style.md

```diff
@@ -0,0 +1,162 @@
+[Spark](https://github.com/apache/spark/blob/a63d9edcfb8a714a17492517927aa114dea8fea0/docs/running-on-mesos.md).
+(Note that the notion of application-specific does sometimes blur into the realm of workload-specific,
+for example [Chronos](https://github.com/mesos/chronos) is a generic framework for batch jobs.
+However, regardless of what set of Mesos frameworks you are using, the key properties remain: each
+framework is its own API endpoint with its own client-facing and internal abstractions, state machines, and scheduler).
+
+A Mesos-style system can avoid two-level scheduling by integrating cluster-level resource allocation logic and
+application-level work allocation logic in a single framework.
+
+## Building a Mesos-style framework on Kubernetes
+
+Implementing the Mesos model on Kubernetes boils down to enabling application-specific collection abstractions,
+controllers/state machines, and scheduling. There are just three steps:
+* Use API plugins to create API resources for your new application-specific collection abstraction(s)
+* Implement controllers for the new abstractions (and for managing the lifecycle of the pods the controllers generate)
+* Implement a scheduler with the application-specific scheduling logic
+
+Note that the last two can be combined: a Kubernetes controller can do the scheduling for the pods it creates,
+by writing `nodeName` to the pods when it creates them.
+
+Once you've done this, you end up with an architecture that is extremely similar to the Mesos-style--the
+Kubernetes controller is effectively a Mesos framework. The remaining differences are
+* In Kubernetes, all API operations go through a single logical endpoint, the API server (we say logical because the API server can be replicated).
+In contrast, in Mesos, API operations go to a particular framework. However, the Kubernetes API plugin model makes this difference fairly small.
+* In Kubernetes, application-specific admission control, quota, defaulting, etc. rules can be implemented
+in the API server rather than in the controller. Of course you can choose to make these operations be no-ops for
+your application-specific collection abstractions, and handle them in your controller.
+* On the node level, Mesos allows application-specific executors, whereas Kubernetes only has
+executors for Docker and Rocket containers.
+
+The end-to-end flow is
+
+1. Client submits an application-specific collection object to the API server
+2. The API server plugin for that collection object forwards the request to the API server that handles that collection type
+3. Admission control, quota, applying defaults, etc. runs on the collection object
+4. If the collection is admitted, it is persisted
+5. The collection controller sees the collection object and in response creates the underlying pods and chooses which nodes they will run on by setting `nodeName`
+6. Kubelet sees the pods with `nodeName` set and starts the container(s)
```

Comment:
The `kubelet` may reject a pod that has been scheduled to it.  I am not sure if this is worth calling out and if the node agent in Mesos has a similar concept?